hepcrawl/spiders/hindawi_spider.py

```diff
@@ -0,0 +1,237 @@
+
+    1. The spider will parse the local MARC21XML format file for record data
+
+    2. Finally a HEPRecord will be created.
+
+
+    Example usage:
+    .. code-block:: console
+
+        scrapy crawl hindawi -a source_file=file://`pwd`/tests/responses/hindawi/test_1.xml
+
+    Happy crawling!
+    """
+
+    name = 'hindawi'
+    start_urls = []
+    iterator = 'xml'
+    itertag = 'marc:record'
+
+    namespaces = [
+        ("OAI-PMH", "http://www.openarchives.org/OAI/2.0/"),
+        ("marc", "http://www.loc.gov/MARC21/slim"),
+        ("mml", "http://www.w3.org/1998/Math/MathML"),
+    ]
+
+    def __init__(self, source_file=None, *args, **kwargs):
+        """Construct Hindawi spider."""
+        super(HindawiSpider, self).__init__(*args, **kwargs)
+        self.source_file = source_file
+
+    def start_requests(self):
+        """Default starting point for scraping shall be the local XML file."""
+        yield Request(self.source_file)
+
+    @staticmethod
+    def get_affiliations(author):
+        """Get the affiliations of an author."""
+        affiliations_raw = author.xpath(
+            "./subfield[@code='u']/text()").extract()
+        affiliations = []
+        for aff in affiliations_raw:
+            affiliations.append(
+                {"value": aff}
+            )
+
+        return affiliations
+
+    def get_authors(self, node):
+        """Gets the authors."""
+        authors_first = node.xpath("./datafield[@tag='100']")
+        authors_others = node.xpath("./datafield[@tag='700']")
+        authors_raw = authors_first + authors_others
+        authors = []
+        for author in authors_raw:
+            authors.append({
+                'raw_name': author.xpath("./subfield[@code='a']/text()").extract_first(),
+                'affiliations': self.get_affiliations(author)
+            })
+
+        return authors
+
+    def get_urls_in_record(self, node):
+        """Return all the different urls in the xml."""
+        marc_856 = node.xpath(
+            "./datafield[@tag='856']/subfield[@code='u']/text()").extract()
+        marc_FFT = node.xpath(
+            "./datafield[@tag='FFT']/subfield[@code='a']/text()").extract()
+        all_links = list(set(marc_856 + marc_FFT))
+
+        return self.differentiate_urls(all_links)
+
+    @staticmethod
+    def differentiate_urls(urls_in_record):
+        """Determine what kind of urls the record has."""
+        pdf_links = []
+        xml_links = []
+        splash_links = []
+        for link in urls_in_record:
+            if "pdf" in link.lower():
+                pdf_links.append(link)
+            elif "xml" in link.lower():
+                xml_links.append(link)
+            elif "dx.doi.org" in link.lower():
+                splash_links.append(link)
+
+        return (
+            pdf_links,
+            xml_links,
+            splash_links,
+        )
+
+    @staticmethod
+    def get_abstract(node):
+        """Get abstract."""
+        return node.xpath(
```

Comment:
Any specific reason this is not directly a `record.xpath` instead of `add_value` since it is a one line xpath expression?